modules/pel/peltool/user_data.py

```diff
@@ -0,0 +1,81 @@
+from pel.datastream import DataStream
+from collections import OrderedDict
+from pel_values import *
+from enum import Enum, unique
+import json
+
+
+@unique
+class UserDataFormat(Enum):
+    json = 0x1
+    cbor = 0x2
+    text = 0x3
+    custom = 0x4
+
+
+def get_value(data: memoryview, start: int, end: int) -> int:
+    return int.from_bytes(data[start: start + end], byteorder="big")
+
+
+class UserData:
+    """
+    This represents the User Data section in a PEL.  It is free form data
+    that the creator knows the contents of.  The component ID, version,
+    and sub-type fields in the section header are used to identify the
+    format.
+
+    The Section base class handles the section header structure that every
+    PEL section has at offset zero.
+    """
+
+    def __init__(self, stream: DataStream, sectionID: int, sectionLen: int,
+                 versionID: int, subType: int, componentID: int, creatorID: str):
+        self.stream = stream
+        self.sectionID = sectionID
+        self.sectionLen = sectionLen
+        self.versionID = versionID
+        self.subType = subType
+        self.componentID = componentID
+        self.creatorID = creatorID
+        self.dataLength = sectionLen - 8
+        self.data = self.stream.get_mem(self.dataLength)
+
+    def paser(self) -> str:
+        import importlib
+        name = (self.creatorID.lower() + "%04X" % self.componentID).lower()
+        try:
+            cls = importlib.import_module("udparsers." + name + "." + name)
+            mv = memoryview(self.data)
+            return cls.parseUDToJson(self.subType, self.versionID, mv)
+        except:
+            print("Failed to parse user data, module = " + name)
+            exit(1)
+
+    def getBuiltinFormatJSON(self, out: OrderedDict):
+        if self.subType == UserDataFormat.json.value:
+            out.update(json.loads(bytes.decode(
+                self.stream.get_mem(self.dataLength)).strip()))
```

Comment:
this crashes because self.stream's index was already incremented to the end by line 41.  You probably want to operate on self.data instead?

Reply:
> this crashes because self.stream's index was already incremented to the end by line 41. You probably want to operate on self.data instead?

Yes, You are right.
changed stream.data